sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala

```diff
@@ -232,7 +232,9 @@ private[csv] object CSVTypeCast {
       nullable: Boolean = true,
       options: CSVOptions = CSVOptions()): Any = {
 
-    if (nullable && datum == options.nullValue) {
+    // datum can be null if the number of fields found is less than the length of the schema
+    if (datum == options.nullValue || datum == null) {
+      require(nullable, "null value found but the field is not nullable.")
```

Comment:
@rxin well, the user input is the illegal argument I guess, but I don't feel strongly about it. I was really arguing against `RuntimeException` which is never really the right thing to throw. `NullPointerException`? `IllegalStateException`? take your pick of anything else more specific.